### ISSUE TITLE
Some questions about the "dynamiclinks" sample...

### ISSUE THREAD

**Issue by AndroidDeveloperLB** [firebase/quickstart-android#111]
I have a few questions about this sample:

1. Suppose I have a "google-services.json" file and an app code from the Firebase console (but of another app, of course), what would it take to check the sample?

  Is it only for checking out how it works?

2. Shouldn't you call "disconnect" on "mGoogleApiClient" variable, once it's not needed anymore?

3. Shouldn't I perform all the deep linking handling (from "START build_api_client" to "END get_deep_link") only once, so that I should use a SharedPreference to avoid checking it again and again, for each run of the app?

4. Suppose I use app invite, would this work fine with dynamic-links:

```
public Intent getAppInviteIntent(Context context) {
    return new AppInviteInvitation.IntentBuilder(title)
            .setCustomImage(imageUri)
            .setMessage(message)
            .setCallToActionText(downloadActionText)
            .setGoogleAnalyticsTrackingId(context.getString(R.string.ga_property_id))
            .setDeepLink(buildDeepLink(...)
            .build();
}
```
?
I want to measure how many people got the app-invitation, vs how many installed the app from app-invitation. I hope this is the way to do it.

**Comment by samtstern** [comment]
@AndroidDeveloperLB hey there, I'll try to answer your questions:

1. Not sure exactly what you're asking here.  The main purpose of the code here is produce a minimal, runnable sample demonstrating how Dynamic Links work and how to use them.

2. Since we are using `enableAutoManage()` there is no need to call `connect()` or `disconnect()`
3. The `getInvitation` API will only return success once per link, after that it will 'mark' it as done, like you suggested.  Therefore it's OK to call it on `onCreate`.
4. App Invites and Dynamic Links are compatible, you can think of App Invites as a 'skin' over Dynamic Links.  The email or SMS generated by the invite will contain a Dynamic Link, which your app will receive.

**Comment by AndroidDeveloperLB** [comment]
1. Thing is, how to make the sample work
2. I see. Maybe you should write a comment there. Does it also include error handling (disconnect and free everything, if an error occurred) ? 
3. So it's not needed to save for next runs. How much does it take to handle all the calls till it decides it has already handled the deep link, compared to just using SharedPreference to check if we already did it in the past?
I think using SharedPreference takes less time.
4. But did I do it correctly?

**Comment by samtstern** [comment]
1. Please follow the [README](https://github.com/firebase/quickstart-android/blob/master/dynamiclinks/README.md) to make the sample work.
2.  You can find more documentation on `enableAutoManage()` [here](https://developers.google.com/android/guides/api-client#start_an_automatically_managed_connection)
3. It uses `SharedPreferences` internally, so it's the same speed.
4. It's hard to say why that's not working.  I am going to close this issue, please open a new issue with more information about your iOS issue.

**Comment by AndroidDeveloperLB** [comment]
1. Thanks.
2. I was talking about the sample, but thank you . I will read it. According to it, using enableAutoManage isn't intended to use if I work outside of an Activity, right? So if I don't use it, I should disconnect in all cases, including error.
3. But it still needs to connect to mGoogleApiClient. 
4. What can I do to handle it correctly?

**Comment by samtstern** [comment]
2. Yes it's safe to use `autoManage` even if the Activity might die before the result comes back, this case is handled.  You're right that `enableAutoManage` is only for use in an `Activity` context, in other contexts you are required for calling `connect()` and `disconnect()`.
3. Yes that's true.  If that delay is a problem for you, you can track this yourself as described.
4. There's nothing obviously wrong with the Android code snippet above.

**Comment by AndroidDeveloperLB** [comment]
1+2+3. ok, thanks.

4. Can you please tell me what I should do? Maybe I can send you the keys, so that you could have a look, or something?

**Comment by samtstern** [comment]
@AndroidDeveloperLB please file an issue on [/firebase/quickstart-ios] to troubleshoot your iOS issue.

**Comment by AndroidDeveloperLB** [comment]
@samtstern I'm not IOS developer. The code I wrote is for Android. I wish to know why it doesn't work, as it goes to an error page on IOS.
